curvefs/src/client/rpcclient/task_excutor.h

```diff
@@ -0,0 +1,195 @@
+#include "curvefs/proto/common.pb.h"
+#include "curvefs/src/client/common/common.h"
+#include "curvefs/src/client/rpcclient/metacache.h"
+#include "curvefs/src/client/rpcclient/channel_manager.h"
+
+using ::curve::client::CopysetID;
+using ::curve::client::LogicPoolID;
+using ::curvefs::client::common::MetaserverID;
+using ::curvefs::client::common::MetaServerOpType;
+using ::curvefs::common::PartitionInfo;
+
+namespace curvefs {
+namespace client {
+namespace rpcclient {
+
+struct ExcutorOpt {
+    uint32_t maxRetry = 3;
+    uint64_t retryIntervalUS = 200;
+    uint64_t rpcTimeoutMS = 1000;
+    uint64_t maxRPCTimeoutMS = 64000;
+    uint64_t maxRetrySleepIntervalUS = 64ull * 1000 * 1000;
+    uint64_t minRetryTimesForceTimeoutBackoff = 5;
+    uint64_t maxRetryTimesBeforeConsiderSuspend = 20;
+};
+
+class TaskContext {
+ public:
+    using RpcFunc = std::function<int(
+        uint32_t poolID, uint32_t copysetID, uint32_t partitionID,
+        uint64_t applyIndex, brpc::Channel *channel, brpc::Controller *cntl)>;
+    TaskContext() {}
+    TaskContext(MetaServerOpType type, RpcFunc func, uint32_t fsid = 0,
+                uint32_t inodeid = 0)
+        : optype(type), rpctask(func), fsID(fsid), inodeID(inodeid) {}
+
+    std::string TaskContextStr() {
+        std::ostringstream oss;
+        oss << "{" << optype << ",fsid=" << fsID << ",inodeid=" << inodeID
+            << "}";
+        return oss.str();
+    }
+
+ public:
+    uint64_t rpcTimeoutMs;
+    MetaServerOpType optype;
+    RpcFunc rpctask = nullptr;
+    uint32_t fsID = 0;
+    uint64_t inodeID = 0;
+
+    LogicPoolID poolID;
+    CopysetID copysetID;
+    MetaserverID targetID;
+    uint32_t partitionID;
+    butil::EndPoint targetAddr;
+    uint64_t applyIndex = 0;
+    bool retryDirectly = false;
+
+    uint32_t retryTimes = 0;
+    bool suspend = false;
+};
+
+class TaskExcutor {
+ public:
+    TaskExcutor() {}
+    TaskExcutor(const ExcutorOpt &opt, std::shared_ptr<MetaCache> metaCache,
+                std::shared_ptr<ChannelManager<MetaserverID>> channelManager)
+        : opt_(opt), metaCache_(metaCache), channelManager_(channelManager) {}
+
+    int DoRPCTask(std::shared_ptr<TaskContext> task);
+
+ protected:
+    bool RetryTimeExceed();
+    bool NeedRetry();
+    bool PreProcessBeforeRetry(int retCode);
+    int ExcuteTask();
+    virtual bool GetTarget();
+    void UpdateApplyIndex(LogicPoolID pid, CopysetID copysetId,
+                          uint64_t applyIndex);
+
+    bool OnReturn(int retCode);
+    void OnSuccess();
+    void OnOverLoad();
+    void OnReDirected();
+    void OnCopysetNotExist();
+
+    void RefreshLeader();
+    uint64_t OverLoadBackOff();
+    uint64_t TimeoutBackOff();
+
+ protected:
+    std::shared_ptr<MetaCache> metaCache_;
+    std::shared_ptr<ChannelManager<MetaserverID>> channelManager_;
+    std::shared_ptr<TaskContext> task_;
+    ExcutorOpt opt_;
+};
+
+class CreateDentryExcutor : public TaskExcutor {
```

Comment:
all executors except the last one `CreateInodeExecutor` are complete identical to `TaskExecutor`, so ethier you can directly use TaskExecutor or just let `CreateDentryExecutor` is an alias to `TaskExecutor` by `using CreateDentryExecutor = TaskExecutor`

Reply:
Or, just define a CommonInodeExexutor is enough.

Reply:
Here is to reserve, different types of error handling may be different. But it can also be modified later when needed.
What you said is a good way, I adopt it.